### PR TITLE
[camera] android-rework part 8: Supporting modules for final implementation

### PR DESCRIPTION
**Pull request by BeMacized**
This PR adds the remaining supporting functionality needed to implement previous parts of the android rework in the camera module (which is likely to be part 9).

This is the eighth PR in a series of pull-requests which will gradually introduce changes from PR #3651, making it easier to review the code (as discussed with @stuartmorgan).

~~As of right now, this PR will still be kept in draft mode, as it also contains changes it depends on from previous parts. Once those have been merged and any changes from those have been merged back into this PR, the draft status will be removed.~~

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides]. (Note that unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`. See [plugin_tool format])
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
  - Since this will be a series of pull-requests I will update the version in the last PR of the series.
- [ ] I updated CHANGELOG.md to add a description of the change.
  - Since this will be a series of pull-requests I will update the version in the last PR of the series.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt.
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by mvanbeusekom** [comment]
@googlebot I consent.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraCaptureCallback.java [hunk 1]
```diff
@@ -0,0 +1,142 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraCaptureCallback.java:3]: Should update the license header with the correct information:

```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraCaptureCallback.java [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import android.hardware.camera2.CameraCaptureSession;
+import android.hardware.camera2.CameraCaptureSession.CaptureCallback;
+import android.hardware.camera2.CaptureRequest;
+import android.hardware.camera2.CaptureResult;
+import android.hardware.camera2.TotalCaptureResult;
+import android.util.Log;
+import androidx.annotation.NonNull;
+import io.flutter.plugins.camera.types.CaptureTimeoutsWrapper;
+
+class CameraCaptureCallback extends CaptureCallback {
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraCaptureCallback.java:16]: Class is missing documentation on public members.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraState.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/CameraState.java:3]: Wrong license header:

```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/DartMessenger.java [hunk 4]
```diff
@@ -16,7 +16,7 @@
 import java.util.HashMap;
 import java.util.Map;
 
-class DartMessenger {
+public class DartMessenger {
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/DartMessenger.java:19]: Class is missing documentation for public members.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/ImageSaver.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/ImageSaver.java:3]: Wrong license header:
```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/ImageSaver.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/** Saves a JPEG {@link Image} into the specified {@link File}. */
+public class ImageSaver implements Runnable {
+
+  /** The JPEG image */
+  private final Image mImage;
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/ImageSaver.java:20]: In other classes the member fields are not prefixed with the `m` character, the same style should be used here.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/ImageSaver.java [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+  /** The file we save the image into. */
+  private final File mFile;
+
+  /** Used to report the status of the save action. */
+  private final Callback mCallback;
+
+  ImageSaver(@NonNull Image image, @NonNull File file, @NonNull Callback callback) {
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/ImageSaver.java:28]: Missing documentation.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/ImageSaver.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        try {
+          output.close();
+        } catch (IOException e) {
+          mCallback.onError("cameraAccess", e.getMessage());
+        }
+      }
+    }
+  }
+
+  public interface Callback {
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/ImageSaver.java:60]: Missing documentation on interface and it's members.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactory.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactory.java:3]: // Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactory.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+import io.flutter.plugins.camera.features.flash.FlashFeature;
+import io.flutter.plugins.camera.features.focuspoint.FocusPointFeature;
+import io.flutter.plugins.camera.features.fpsrange.FpsRangeFeature;
+import io.flutter.plugins.camera.features.noisereduction.NoiseReductionFeature;
+import io.flutter.plugins.camera.features.resolution.ResolutionFeature;
+import io.flutter.plugins.camera.features.resolution.ResolutionPreset;
+import io.flutter.plugins.camera.features.sensororientation.SensorOrientationFeature;
+import io.flutter.plugins.camera.features.zoomlevel.ZoomLevelFeature;
+
+public interface CameraFeatureFactory {
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactory.java:24]: Missing documentation

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactoryImpl.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactoryImpl.java:3]: Wrong license header:

```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactoryImpl.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+import io.flutter.plugins.camera.features.flash.FlashFeature;
+import io.flutter.plugins.camera.features.focuspoint.FocusPointFeature;
+import io.flutter.plugins.camera.features.fpsrange.FpsRangeFeature;
+import io.flutter.plugins.camera.features.noisereduction.NoiseReductionFeature;
+import io.flutter.plugins.camera.features.resolution.ResolutionFeature;
+import io.flutter.plugins.camera.features.resolution.ResolutionPreset;
+import io.flutter.plugins.camera.features.sensororientation.SensorOrientationFeature;
+import io.flutter.plugins.camera.features.zoomlevel.ZoomLevelFeature;
+
+public class CameraFeatureFactoryImpl implements CameraFeatureFactory {
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactoryImpl.java:24]: Missing documentation

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatures.java [hunk 13]
```diff
@@ -0,0 +1,35 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatures.java:3]: Wrong license header:

```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/types/CaptureTimeoutsWrapper.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
+package io.flutter.plugins.camera.types;
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/types/CaptureTimeoutsWrapper.java:1]: Missing license header:

```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.camera.types;
```

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/types/CaptureTimeoutsWrapper.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+public class CaptureTimeoutsWrapper {
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/types/CaptureTimeoutsWrapper.java:3]: Missing documentation

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/types/Timeout.java [hunk 16]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/types/Timeout.java:3]: Wrong license header:
```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

File: packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/CameraCaptureCallbackStatesTest.java [hunk 17]
```diff
@@ -0,0 +1,373 @@
+package io.flutter.plugins.camera;
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/CameraCaptureCallbackStatesTest.java:1]: Missing license header:
```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.camera;
```

File: packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/CameraRegionUtilsTest.java [hunk 18]
```diff
@@ -0,0 +1,294 @@
+package io.flutter.plugins.camera;
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/CameraRegionUtilsTest.java:1]: Missing license header:

```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.camera;
```

File: packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/ImageSaverTests.java [hunk 19]
```diff
@@ -0,0 +1,101 @@
+package io.flutter.plugins.camera;
```
Comment (mvanbeusekom) [inline packages/camera/camera/android/src/test/java/io/flutter/plugins/camera/ImageSaverTests.java:1]: Missing license header:

```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.camera;
```

**Review by mvanbeusekom — CHANGES_REQUESTED** [review]
Did a quick scan mainly on documentation and license headers. Lets fix this first and then do a more detailed deep dive.

File: packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactory.java (hunk 9 above)
Comment (BeMacized) [reply packages/camera/camera/android/src/main/java/io/flutter/plugins/camera/features/CameraFeatureFactory.java:3]:
```suggestion
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.
```

**Comment by acoutts** [comment]
@googlebot I consent.

**Review by mvanbeusekom — APPROVED** [review]
LGTM

**Pull request merged**